core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go

```diff
@@ -329,46 +329,88 @@ func (p *logEventProvider) readLogs(ctx context.Context, latest int64, entries [
+		// the block limitation with maxBurst
+		// fmt.Printf("reserving: %d for %s\n", int(latest-start), entry.upkeepID)
 		resv := entry.blockLimiter.ReserveN(time.Now(), int(latest-start))
+		// fmt.Printf("reservation limit for %s: %d\n", entry.upkeepID, resv.DelayFrom(time.Now())/time.Millisecond)
 		if !resv.OK() {
-			merr = multierr.Append(merr, fmt.Errorf("%s: %s", BlockLimitExceeded, entry.upkeepID.String()))
+			// the block range cannot be processed for the event either because
+			// it is above the configured rate or burst limit
+			merr = errors.Join(merr, fmt.Errorf("%w: %s", ErrBlockLimitExceeded, entry.upkeepID.String()))
+
 			continue
 		}
-		start = start - p.opts.LookbackBuffer // adding a buffer to check for reorgs
+
+		// adding a buffer outside the reserved block range to check for reorgs
+		start = start - p.opts.LookbackBuffer
 		if start < 0 {
 			start = 0
 		}
-		// lggr = lggr.With("startBlock", start)
+
 		logs, err := p.poller.LogsWithSigs(start, latest, entry.topics, common.BytesToAddress(entry.addr), pg.WithParentCtx(ctx))
 		if err != nil {
-			resv.Cancel() // cancels limit reservation as we failed to get logs
+			// cancel limit reservation as we failed to get logs
+			resv.Cancel()
+
 			if ctx.Err() != nil {
-				return multierr.Append(merr, ctx.Err())
+				return errors.Join(merr, ctx.Err())
 			}
-			merr = multierr.Append(merr, fmt.Errorf("failed to get logs for upkeep %s: %w", entry.upkeepID.String(), err))
+
+			merr = errors.Join(merr, fmt.Errorf("failed to get logs for upkeep %s: %w", entry.upkeepID.String(), err))
+
+			// continue processing entries as this is not a hard error
 			continue
 		}
-		// if this limiter's burst was set to the max,
-		// we need to reset it
+
+		// if this limiter's burst was set to the max reset it and cancel the
+		// reservation to allow further processing without exceeding the rate
+		// limit
 		if entry.blockLimiter.Burst() == maxBurst {
-			resv.Cancel() // cancel the reservation as we are resetting the burst
+			// cancel the reservation as we are resetting the burst
+			resv.Cancel()
+
 			entry.blockLimiter.SetBurst(p.opts.BlockLimitBurst)
 		}
-		added := p.buffer.enqueue(entry.upkeepID, logs...)
-		// if we added logs or couldn't find, update the last poll block
-		if added > 0 || len(logs) == 0 {
-			entry.lastPollBlock = latest
-		}
-		// if n := len(logs); n > 0 {
-		// 	lggr.Debugw("got logs for upkeep", "logs", n, "added", added)
+
+		// add logs returned from the poller to the queue and return the total
+		// number of unique logs added to the queue
+		// added :=
+		p.buffer.enqueue(entry.upkeepID, logs...)
+
+		// fmt.Printf("added: %d; logs: %d\n", added, len(logs))
+
+		// if no new logs were added or no logs were found, update the last poll block
+		// if added > 0 || len(logs) == 0 {
+		entry.lastPollBlock = latest
+
+		p.filterStore.UpdateFilters(func(original upkeepFilter, updated upkeepFilter) upkeepFilter {
```

Comment:
We update the last poll block from another function

Reply:
In that case, we also don't need `entry.lastPollBlock = latest` in the condition of `added > 0 || len(logs) == 0`?

Reply:
NVM